File: ios/Sources/ACRendering/Views/CompoundButtonView.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import SwiftUI
import ACCore
import ACAccessibility
import ACFluentUI

struct CompoundButtonView: View {
    let button: CompoundButton
    let hostConfig: HostConfig

    @Environment(\.actionHandler) var actionHandler
    @Environment(\.actionDelegate) var actionDelegate
    @EnvironmentObject var viewModel: CardViewModel

    private enum Layout {
        static let iconSize: CGFloat = 24
        static let iconTextSpacing: CGFloat = 12
        static let titleSubtitleSpacing: CGFloat = 4
        static let horizontalPadding: CGFloat = 16
        static let verticalPadding: CGFloat = 12
        static let minHeight: CGFloat = 44
        static let badgeFontSize: CGFloat = 10
    }

    var body: some View {
        Button(action: handleAction) {
            content
        }
        .buttonStyle(CompoundButtonStyle(
            style: button.style ?? "default",
            isDisabled: button.selectAction == nil,
            hostConfig: hostConfig
        ))
        .disabled(button.selectAction == nil)
        .accessibilityLabel(accessibilityLabel)
        .accessibilityHint(accessibilityHint)
    }

    private var isStyledButton: Bool {
        let s = button.style ?? "default"
        return ["emphasis", "positive", "destructive"].contains(s)
    }

    private var primaryTextColor: Color {
        isStyledButton ? .white : Color(hex: hostConfig.containerStyles.default.foregroundColors.default.default)
    }

    private var secondaryTextColor: Color {
        isStyledButton ? .white.opacity(0.8) : Color(hex: hostConfig.containerStyles.default.foregroundColors.default.subtle)
    }

    @ViewBuilder
    private var content: some View {
        HStack(alignment: .center, spacing: Layout.iconTextSpacing) {
            if button.iconPosition != "trailing" {
                iconView
            }

            VStack(alignment: .leading, spacing: Layout.titleSubtitleSpacing) {
                HStack {
                    Text(button.title)
                        .font(.system(size: CGFloat(hostConfig.fontSizes.large), weight: titleFontWeight))
                        .foregroundColor(primaryTextColor)
                        .lineLimit(2)
                        .truncationMode(.tail)

                    if let badge = button.badge {
                        Text(badge)
                            .font(.system(size: Layout.badgeFontSize, weight: .semibold))
                            .foregroundColor(.white)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color(hex: hostConfig.containerStyles.default.foregroundColors.accent.`default`))
                            .cornerRadius(4)
                            .lineLimit(1)
                    }
                }

                if let description = button.description {
                    Text(description)
                        .font(.system(size: CGFloat(hostConfig.fontSizes.default)))
                        .foregroundColor(secondaryTextColor)
                        .lineLimit(2)
                        .truncationMode(.tail)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            if button.iconPosition == "trailing" {
                iconView
            }

            // Chevron indicator
            Image(systemName: "chevron.right")
                .font(.system(size: CGFloat(hostConfig.fontSizes.default), weight: titleFontWeight))
                .foregroundColor(secondaryTextColor)
        }
        .padding(.horizontal, Layout.horizontalPadding)
        .padding(.vertical, Layout.verticalPadding)
        .frame(minHeight: Layout.minHeight)
    }

    private var titleFontWeight: Font.Weight {
        let weightValue = hostConfig.fontWeights.bolder
        switch weightValue {
        case 100...199: return .ultraLight
        case 200...299: return .light
        case 300...399: return .regular
        case 400...499: return .regular
        case 500...599: return .medium
        case 600...699: return .semibold
        case 700...799: return .bold
        default: return .heavy
        }
    }

    @ViewBuilder
    private var iconView: some View {
        if let iconName = button.iconName {
            if iconName.hasPrefix("http://") || iconName.hasPrefix("https://") {
                AsyncImage(url: URL(string: iconName)) { phase in
                    switch phase {
                    case .success(let image):
                        image
                            .resizable()
                            .scaledToFit()
                            .frame(width: Layout.iconSize, height: Layout.iconSize)
                    case .failure:
                        iconPlaceholder
                    case .empty:
                        ProgressView()
                            .frame(width: Layout.iconSize, height: Layout.iconSize)
                    @unknown default:
                        iconPlaceholder
                    }
                }
            } else {
                // Map Fluent icon names to SF Symbols, then fall back to raw name
                let sfSymbol = Self.resolveFluentIcon(iconName)
                Image(systemName: sfSymbol)
                    .resizable()
                    .scaledToFit()
                    .frame(width: Layout.iconSize, height: Layout.iconSize)
                    .foregroundColor(primaryTextColor)
            }
        }
    }

    /// Maps Fluent UI icon names to SF Symbol equivalents.
    private static func resolveFluentIcon(_ name: String) -> String {
        let baseName = name
            .replacingOccurrences(of: ",Filled", with: "")
            .replacingOccurrences(of: ",Regular", with: "")
            .trimmingCharacters(in: .whitespaces)
        let lookup: [String: String] = [
            "Calendar": "calendar",
            "Info": "info.circle",
            "InfoCircle": "info.circle",
            "Warning": "exclamationmark.triangle",
            "Error": "xmark.circle",
            "Checkmark": "checkmark.circle",
            "CheckmarkCircle": "checkmark.circle",
            "Search": "magnifyingglass",
            "Settings": "gearshape",
            "Person": "person",
            "People": "person.2",
            "Mail": "envelope",
            "Chat": "bubble.left",
            "Phone": "phone",
            "Video": "video",
            "Camera": "camera",
            "Document": "doc",
            "Folder": "folder",
            "Star": "star",
            "Heart": "heart",
            "Home": "house",
            "Location": "location",
            "Map": "map",
            "Clock": "clock",
            "Alert": "bell",
            "Add": "plus.circle",
            "Delete": "trash",
            "Edit": "pencil",
            "Share": "square.and.arrow.up",
            "Link": "link",
            "Globe": "globe",
            "Lock": "lock",
            "ArrowRight": "arrow.right",
            "ChevronRight": "chevron.right",
            "Dismiss": "xmark",
            "MoreHorizontal": "ellipsis",
            "Attach": "paperclip",
            "Send": "paperplane",
            "Airplane": "airplane",
            "Food": "fork.knife",
            "Gift": "gift",
            "Money": "dollarsign.circle",
            "Weather": "cloud.sun",
            "Flash": "bolt",
            "Play": "play",
            "Mic": "mic",
        ]
        return lookup[baseName] ?? name
    }

    private var iconPlaceholder: some View {
        Image(systemName: "photo")
            .resizable()
            .scaledToFit()
            .frame(width: Layout.iconSize, height: Layout.iconSize)
            .foregroundColor(Color(hex: hostConfig.containerStyles.default.foregroundColors.default.subtle))
    }

    private var accessibilityLabel: String {
        if let description = button.description {
            return "\(button.title). \(description)"
        }
        return button.title
    }

    private var accessibilityHint: String {
        guard let action = button.selectAction else {
            return ""
        }

        switch action {
        case .openUrl:
            return "Opens URL"
        case .submit:
            return "Submits form"
        case .showCard:
            return "Shows card"
        case .toggleVisibility:
            return "Toggles visibility"
        case .execute:
            return "Executes action"
        case .popover:
            return "Shows popover"
        case .runCommands:
            return "Runs commands"
        case .openUrlDialog:
            return "Opens URL dialog"
        case .unknown:
            return ""
        }
    }

    private func handleAction() {
        guard let action = button.selectAction else { return }
        actionHandler.handle(action, delegate: actionDelegate, viewModel: viewModel)
    }
}

// Custom button style using hostConfig for theming
struct CompoundButtonStyle: ButtonStyle {
    let style: String
    let isDisabled: Bool
    let hostConfig: HostConfig

    func makeBody(configuration: Configuration) -> some View {
        let cornerRadius = CGFloat(hostConfig.cornerRadius["container"] ?? 4)

        configuration.label
            .background(backgroundColor.opacity(configuration.isPressed ? 0.8 : 1.0))
            .cornerRadius(cornerRadius)
            .shadow(
                color: Color.black.opacity(isDisabled ? 0 : 0.1),
                radius: 2,
                x: 0,
                y: 1
            )
            .opacity(isDisabled ? 0.5 : 1.0)
            .overlay(
                RoundedRectangle(cornerRadius: cornerRadius)
                    .stroke(Color(hex: hostConfig.compoundButton.borderColor), lineWidth: CGFloat(hostConfig.separator.lineThickness))
            )
    }

    private var isStyledButton: Bool {
        ["emphasis", "positive", "destructive"].contains(style)
    }

    private var backgroundColor: Color {
        let colors = hostConfig.containerStyles.default.foregroundColors
        switch style {
        case "emphasis":
            return Color(hex: colors.accent.`default`)
        case "positive":
            return Color(hex: colors.good.`default`)
        case "destructive":
            return Color(hex: colors.attention.`default`)
        default:
            return Color(hex: hostConfig.containerStyles.default.backgroundColor)
        }
    }
}

File: android/ac-rendering/src/main/kotlin/com/microsoft/adaptivecards/rendering/composables/CompoundButtonView.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.rendering.composables

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import coil.compose.AsyncImage
import coil.request.ImageRequest
import com.microsoft.adaptivecards.core.models.CompoundButton
import com.microsoft.adaptivecards.rendering.theme.LocalHostConfig
import com.microsoft.adaptivecards.rendering.viewmodel.ActionHandler

// Layout constants for consistency
private object CompoundButtonLayout {
    val IconSize = 24.dp
    val IconTextSpacing = 12.dp
    val TitleSubtitleSpacing = 4.dp
    val HorizontalPadding = 16.dp
    val VerticalPadding = 12.dp
    val CornerRadius = 8.dp
    val MinHeight = 48.dp
    val Elevation = 2.dp
    val BadgeFontSize = 10.sp
}

/**
 * Renders a CompoundButton element with icon, title, description, and badge
 */
@OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)
@Composable
fun CompoundButtonView(
    element: CompoundButton,
    actionHandler: ActionHandler,
    modifier: Modifier = Modifier
) {
    val hostConfig = LocalHostConfig.current

    // Use hostConfig containerStyles to match iOS styling (subtle fills, not solid blocks)
    val containerColor = when (element.style) {
        "emphasis" -> parseHostColor(hostConfig.containerStyles.accent.backgroundColor)
            ?: MaterialTheme.colorScheme.primaryContainer
        "positive" -> parseHostColor(hostConfig.containerStyles.good.backgroundColor)
            ?: Color(0xFFD5F0DD)
        "destructive" -> parseHostColor(hostConfig.containerStyles.attention.backgroundColor)
            ?: Color(0xFFF7E9E9)
        else -> parseHostColor(hostConfig.containerStyles.default.backgroundColor)
            ?: MaterialTheme.colorScheme.surface
    }

    val contentColor = when (element.style) {
        "emphasis" -> parseHostColor(hostConfig.containerStyles.accent.foregroundColors.default.default)
            ?: MaterialTheme.colorScheme.onPrimaryContainer
        "positive" -> parseHostColor(hostConfig.containerStyles.good.foregroundColors.default.default)
            ?: MaterialTheme.colorScheme.onSurface
        "destructive" -> parseHostColor(hostConfig.containerStyles.attention.foregroundColors.default.default)
            ?: MaterialTheme.colorScheme.onSurface
        else -> MaterialTheme.colorScheme.onSurface
    }

    val isEnabled = element.selectAction != null

    val contentDesc = buildString {
        append(element.title)
        element.description?.let { append(". $it") }
    }

    Card(
        onClick = {
            element.selectAction?.let { action ->
                when (action) {
                    is com.microsoft.adaptivecards.core.models.ActionOpenUrl -> actionHandler.onOpenUrl(action.url, action.id)
                    is com.microsoft.adaptivecards.core.models.ActionSubmit -> actionHandler.onSubmit(emptyMap(), action.id)
                    is com.microsoft.adaptivecards.core.models.ActionExecute -> actionHandler.onExecute(action.verb ?: "", emptyMap(), action.id)
                    else -> { /* Other action types */ }
                }
            }
        },
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = CompoundButtonLayout.MinHeight)
            .semantics { contentDescription = contentDesc },
        colors = CardDefaults.cardColors(
            containerColor = containerColor,
            contentColor = contentColor
        ),
        elevation = CardDefaults.cardElevation(
            defaultElevation = if (element.style == "default") CompoundButtonLayout.Elevation else 0.dp
        ),
        shape = MaterialTheme.shapes.medium,
        enabled = isEnabled
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    horizontal = CompoundButtonLayout.HorizontalPadding,
                    vertical = CompoundButtonLayout.VerticalPadding
                ),
            horizontalArrangement = Arrangement.spacedBy(CompoundButtonLayout.IconTextSpacing),
            verticalAlignment = Alignment.CenterVertically
        ) {
            // Leading icon
            if (element.iconPosition != "trailing") {
                IconView(element.iconName, contentColor)
            }

            // Title, badge, and description
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(CompoundButtonLayout.TitleSubtitleSpacing)
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(
                        text = element.title,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = contentColor,
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.weight(1f, fill = false)
                    )

                    element.badge?.let { badge ->
                        val badgeColor = parseHostColor(
                            hostConfig.containerStyles.default.foregroundColors.accent.default
                        ) ?: MaterialTheme.colorScheme.primary
                        Text(
                            text = badge,
                            fontSize = CompoundButtonLayout.BadgeFontSize,
                            fontWeight = FontWeight.SemiBold,
                            color = Color.White,
                            maxLines = 1,
                            modifier = Modifier
                                .background(
                                    badgeColor,
                                    RoundedCornerShape(4.dp)
                                )
                                .padding(horizontal = 6.dp, vertical = 2.dp)
                        )
                    }
                }

                element.description?.let { description ->
                    Text(
                        text = description,
                        fontSize = 14.sp,
                        color = contentColor.copy(alpha = 0.7f),
                        maxLines = 2,
                        overflow = TextOverflow.Ellipsis
                    )
                }
            }

            // Trailing icon
            if (element.iconPosition == "trailing") {
                IconView(element.iconName, contentColor)
            }

            // Chevron indicator (matching iOS chevron.right style)
            Icon(
                imageVector = Icons.AutoMirrored.Filled.KeyboardArrowRight,
                contentDescription = null,
                modifier = Modifier.size(20.dp),
                tint = contentColor.copy(alpha = 0.6f)
            )
        }
    }
}

@Composable
private fun IconView(iconString: String?, tintColor: Color) {
    if (iconString == null) return

    if (iconString.startsWith("http://") || iconString.startsWith("https://")) {
        AsyncImage(
            model = ImageRequest.Builder(LocalContext.current)
                .data(iconString)
                .crossfade(true)
                .build(),
            contentDescription = null,
            modifier = Modifier.size(CompoundButtonLayout.IconSize),
            contentScale = ContentScale.Fit,
            error = painterResource(android.R.drawable.ic_menu_gallery),
            placeholder = painterResource(android.R.drawable.ic_menu_gallery)
        )
    } else {
        // Resolve Fluent icon name to Material filled icon (matching iOS SF Symbol style)
        val resolved = resolveCompoundButtonIcon(iconString)
        Icon(
            imageVector = resolved,
            contentDescription = null,
            modifier = Modifier.size(CompoundButtonLayout.IconSize),
            tint = tintColor
        )
    }
}

private fun resolveCompoundButtonIcon(name: String): androidx.compose.ui.graphics.vector.ImageVector {
    val lookup = name.split(",").firstOrNull()?.lowercase() ?: name.lowercase()
    return when (lookup) {
        "calendar" -> Icons.Filled.DateRange
        "send" -> Icons.Filled.Send
        "edit" -> Icons.Filled.Edit
        "delete" -> Icons.Filled.Delete
        "add" -> Icons.Filled.Add
        "search" -> Icons.Filled.Search
        "share" -> Icons.Filled.Share
        "star" -> Icons.Filled.Star
        "heart" -> Icons.Filled.Favorite
        "bookmark" -> Icons.Filled.Bookmark
        "info" -> Icons.Filled.Info
        "warning" -> Icons.Filled.Warning
        "settings" -> Icons.Filled.Settings
        "mail", "email" -> Icons.Filled.Email
        "phone", "call" -> Icons.Filled.Call
        "location" -> Icons.Filled.LocationOn
        "person", "peopleteam" -> Icons.Filled.Person
        "home" -> Icons.Filled.Home
        "notification", "bell", "alert" -> Icons.Filled.Notifications
        "check", "checkmark" -> Icons.Filled.Check
        "close", "dismiss" -> Icons.Filled.Close
        else -> Icons.Filled.Info
    }
}

private fun parseHostColor(hex: String?): Color? {
    if (hex == null) return null
    return try {
        Color(android.graphics.Color.parseColor(hex))
    } catch (_: Exception) {
        null
    }
}
